test_conformance/extensions/cl_ext_buffer_device_address/buffer_device_address.cpp

```diff
@@ -435,3 +440,172 @@ REGISTER_TEST(private_address)
+
+        error = clSetKernelArgDevicePointerEXT(kernel, 0, addresses[i] + 8);
+        test_error_ret(error, "Unable to set kernel arg\n", TEST_FAIL);
+
+        const cl_ulong pattern = 0xAABBCCDDEEFF0011 + i;
+        error = clSetKernelArg(kernel, 1, sizeof(pattern), &pattern);
+        test_error_ret(error, "Unable to set kernel arg\n", TEST_FAIL);
+
+        size_t gwo = 0;
+        size_t gws = 1;
+        size_t lws = 1;
+        error = clEnqueueNDRangeKernel(queue, kernel, 1, &gwo, &gws, &lws, 0,
+                                       nullptr, nullptr);
+        test_error_ret(error, "Unable to enqueue kernel\n", TEST_FAIL);
+
+        error = clFinish(queue);
+        test_error_ret(error, "clFinish failed\n", TEST_FAIL);
+
+        std::vector<cl_ulong> results(2, 0);
+        error = clEnqueueReadBuffer(queue, buffer, CL_BLOCKING, 0,
+                                    results.size()
+                                        * sizeof(cl_ulong),
+                                    results.data(), 0, nullptr, nullptr);
+        test_error_ret(error, "clEnqueueReadBuffer failed\n", TEST_FAIL);
+
+        if (results[1] != pattern)
+            test_fail("Test value doesn't match expected value\n");
+    }
+    return TEST_PASS;
+}
+
+REGISTER_TEST(negative_private_address)
+{
+    REQUIRE_EXTENSION("cl_ext_buffer_device_address");
+
+    cl_int error = CL_SUCCESS;
+
+    GET_PFN(device, clSetKernelArgDevicePointerEXT);
+
+    /* Create buffer */
+    clMemWrapper buffer = clCreateBufferWithProperties(
+        context, nullptr, CL_MEM_READ_WRITE, 16, nullptr, &error);
+    cl_mem_device_address_ext address;
+    error = clGetMemObjectInfo(buffer, CL_MEM_DEVICE_ADDRESS_EXT,
+                               sizeof(cl_mem_device_address_ext), &address,
+                               nullptr);
+    test_failure_error_ret(
+        error, CL_INVALID_OPERATION,
+        "clGetMemObjectInfo should return CL_INVALID_OPERATION when: "
+        "\"the buffer was not created with CL_MEM_DEVICE_PRIVATE_ADDRESS_EXT\"",
+        TEST_FAIL);
+
+    static std::string source = R"(
+        void kernel test_device_address(
+            global ulong* ptr,
+            ulong value)
+        {
+            *ptr = value;
+        })";
+
+    clProgramWrapper program;
+    clKernelWrapper kernel;
+    const char *source_ptr = source.data();
+    error = create_single_kernel_helper(context, &program, &kernel, 1,
+                                        &source_ptr, "test_device_address");
+    test_error(error, "Unable to create test kernel");
+
+    error = clSetKernelArgDevicePointerEXT(nullptr, 0, 0);
+    test_failure_error_ret(
+        error, CL_INVALID_KERNEL,
+        "clSetKernelArgDevicePointerEXT should return CL_INVALID_KERNEL when: "
+        "\"kernel is not a valid kernel object\"",
+        TEST_FAIL);
+
+    error = clSetKernelArgDevicePointerEXT(kernel, 1, 0x15465);
```

Comment:
As per the spec "The device pointer can only be used for arguments that are declared to be a pointer to global memory allocated with clCreateBufferWithProperties with the CL_MEM_DEVICE_PRIVATE_ADDRESS_EXT property" 
So this is indeed an error condition but the spec does not mention the error to be returned in this case. This needs to be discussed and clarified in the spec.

Reply:
https://github.com/KhronosGroup/OpenCL-Docs/pull/1492